### PR TITLE
cli(init): webpack4 ready

### LINKED ISSUE #350: `webpack-cli init`  - Support for webpack version 4
**Do you want to request a *feature* or report a *bug*?**

 - Feature
**What is the current behavior?**
Creates an old starter template for webpack.config.js
**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
CommonsChunkPlugin is deprecated and would not expect it in the template.
Add a question which version of webpack the user wants to use.

### PR DESCRIPTION
**Pull request by EugeneHlushko**
**What kind of change does this PR introduce?**

Make `init` command generate webpack4 ready config file

**Did you add tests for your changes?**
no

**If relevant, did you update the documentation?**
no

**Summary**
Fixes #350 

**Does this PR introduce a breaking change?**

**Other information**

### CODE REVIEW

File: lib/generators/init-generator.js [hunk 1]
```diff
@@ -339,8 +338,7 @@ module.exports = class InitGenerator extends Generator {
 				if (regExpForStyles) {
 					if (this.isProd) {
 						this.configuration.config.topScope.push(tooltip.cssPlugin());
-						// TODO: Replace with regular version once v.4 is out
-						this.dependencies.push("extract-text-webpack-plugin@next");
+						this.dependencies.push("extract-text-webpack-plugin");
```
Comment (dhruvdutt) [inline lib/generators/init-generator.js]: We'll still need to use the pre-release version.

Comment (EugeneHlushko) [reply lib/generators/init-generator.js]: yup just pushed smth to reference issue

Comment (EugeneHlushko) [reply lib/generators/init-generator.js]: will work on it tonight :)

**Comment by EugeneHlushko** [comment]
Yes it does, do you want to keep it simple like this? @ev1stensberg 
```
optimization.splitChunks: {
  chunks: "all",
  minChunks: 2
}
```

**Comment by ev1stensberg** [comment]
Go for it

**Comment by ev1stensberg** [comment]
Keep the commonChunk tip, but explain what `splItChunks` is, and that they can remove that prop if they want to.

File: lib/generators/utils/tooltip.js [hunk 2]
```diff
@@ -17,13 +17,18 @@ module.exports = {
...
+ * SplitChunksPlugin is enabled by default and replaced
+ * deprecated CommonsChunkPlugin. It automatically identifies modules which
+ * should be splitted of chunk by heuristics using module duplication count and
+ * module category (i. e. node_modules). And splits the chunks…
  *
- * https://webpack.js.org/plugins/commons-chunk-plugin/
+ * It is safe to remove "splitChunks" from the generated configuration
+ * and was added as an educational example.
+ *
+ * Documentation page will be available here soon:
```
Comment (dhruvdutt) [inline lib/generators/utils/tooltip.js]: We will need to rephrase this a little once [#1863](https://github.com/webpack/webpack.js.org/pull/1863) gets merged.

**Comment by EugeneHlushko** [comment]
Sorry for late notice, just realized i mixed up `minChunks` with `minSize` for `init` generator :)

**Comment by EugeneHlushko** [comment]
Sorry but i didn't quite get that @ev1stensberg 
Did you mean adding a caching group per entry ?

**Comment by ev1stensberg** [comment]
yep

File: lib/generators/utils/entry.js [hunk 3]
```diff
@@ -55,7 +55,7 @@ module.exports = (self, answer) => {
 					self.prompt([
 						InputValidate(
 							`${entryProp}`,
-							`What is the location of "${entryProp}"? [example: "./src/${entryProp}"]`,
+							`What is the location of "${entryProp}"? [example: './src/${entryProp}']`,
```
Comment (EugeneHlushko) [inline lib/generators/utils/entry.js]: This one was showing example in double quotes when other lines use single quotes, changed it. If there was no good reason for it?

File: lib/generators/init-generator.js [hunk 4]
```diff
@@ -375,14 +375,39 @@ module.exports = class InitGenerator extends Generator {
...
+				if (entry && typeof entry === "object") {
+					// if user had provided multiple entry points, we make cacheGroup example
+					// per entry point
+					const entries = Object.keys(entry);
+					if (entries.length > 1) {
+						this.configuration.config.webpackOptions.optimization.splitChunks.cacheGroups = {};
+						entries.map(item => {
+							const newCacheGroup = {
+								chunks: "'all'",
+								// TODO: implement RegExp from entry for test value of caching group
```
Comment (EugeneHlushko) [inline lib/generators/init-generator.js]: Implementing a RegExp based on entry is not working for most cases provided for entry of the example config. It might need work on entry answer parser and optimize transformer. Right now regexp if done similar to scss ones outputs bad examples e.g.
\".\/src\/app'.js
Could use `path` do detect delimiters and `dirname` of entry and then to try putting up a regexp close to what we have in the docs: https://github.com/webpack/webpack.js.org/pull/1863/files#diff-e08b426c523b9305ce571576d5b640d0R165

But then the default scenario when you have `./src/app` and `./src/someOtherEntry` would both result in `./src/app.js` and `./src/someOtherEntry.js` and both dirnames would come put as `src` so that regexp for `cacheGroups[group].test` would be the same `\src\`

**Comment by EugeneHlushko** [comment]
@ev1stensberg i did put up an update but its not very satisfying experience for simple examples like `./src/app` and `./src/otherApp` etc. Pls have a look

File: lib/generators/utils/tooltip.js (hunk 2 above)
Comment (dhruvdutt) [reply lib/generators/utils/tooltip.js]: The documentation is available now. Can you rephrase this?

**Comment by EugeneHlushko** [comment]
Hi @ev1stensberg 
Did you see my previous comment on this? I can resolve the conflict and apply those changes i suggest vs implementing a smelly RegExp for dynamic caching group as this is done for **example** anyway.

File: lib/generators/utils/tooltip.js (hunk 2 above)
Comment (EugeneHlushko) [reply lib/generators/utils/tooltip.js]: webpack/webpack.js.org#1955 was merged, I'll rephrase this when resolving conflict

**Comment by ev1stensberg** [comment]
Implement the one that you think is the best, I'll re-review once done

**Comment by EugeneHlushko** [comment]
@ev1stensberg pls do now

File: lib/generators/init-generator.js [hunk 5]
```diff
@@ -378,15 +374,28 @@ module.exports = class InitGenerator extends Generator {
...
-					oneOrMoreEntries.forEach(prop => {
-						this.configuration.config.webpackOptions.plugins.push(
-							createCommonsChunkPlugin(prop)
-						);
-					});
-				}
+				// add splitChunks options for transparency
+				this.configuration.config.topScope.push(tooltip.splitChunks());
+				this.configuration.config.webpackOptions.optimization = {
+					splitChunks: {
```
[the remark this answers is not in the document]
Comment (EugeneHlushko) [reply lib/generators/init-generator.js]: they are defaults from the docs, we also add a comment to that link here:
https://github.com/webpack/webpack-cli/pull/356/files#diff-f722437ab3fa8013b0afe4475d557c7fR27

and here is direct link:  https://webpack.js.org/plugins/split-chunks-plugin/#optimization-splitchunks

[the remark this answers is not in the document]
Comment (ematipico) [reply lib/generators/init-generator.js]: Oh yeah, I meant to ask if you put these links INTO the code as a comment :)
Sorry for misunderstanding

[the remark this answers is not in the document]
Comment (EugeneHlushko) [reply lib/generators/init-generator.js]: Yeah as a quick ref 👍

File: lib/generators/init-generator.js [hunk 6]
```diff
@@ -378,15 +374,29 @@ module.exports = class InitGenerator extends Generator {
...
+						minChunks: 1,
+						// for production name is recommended to be off
+						name: !this.isProd,
+						cacheGroups: {
+							vendors: {
+								test: "/[\\\\/]node_modules[\\\\/]/",
+								priority: -10
+							},
+						},
+						default: {
```
Comment (ev1stensberg) [inline lib/generators/init-generator.js]: This didn't work when trying to build a web app using this PR. Removing `default` makes the app run. I think maybe that you can use the object entries from when the user types what the names of their entries should be to add chunkGroups. Say if I answer yes to the first question at the CLI and type that I've got 2 entries, `app` and `utils`, then you should use that when setting chunk groups in addition to the node_modules for vendor. Makes sense?

File: lib/generators/utils/entry.js [hunk 7]
```diff
@@ -20,7 +20,7 @@ module.exports = (self, answer) => {
 			.prompt([
 				InputValidate(
 					"multipleEntries",
-					"Type the names you want for your modules (entry files), separated by comma [example: 'app,vendor']",
+					"Type the names you want for your modules (entry files), separated by comma [example: app,vendor]",
```
Comment (ev1stensberg) [inline lib/generators/utils/entry.js]: Use this to split groups of chunks.

File: lib/generators/utils/tooltip.js [hunk 8]
```diff
@@ -17,13 +17,17 @@ module.exports = {
  *
  */`;
 	},
-	commonsChunk: _ => {
+	splitChunks: _ => {
```
Comment (ev1stensberg) [inline lib/generators/utils/tooltip.js]: Good stuff :3

File: lib/init/transformations/optimization/optimization.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+"use strict";
+
+const utils = require("../../../utils/ast-utils");
```
Comment (ev1stensberg) [inline lib/init/transformations/optimization/optimization.js]: This is okay for now, GSOC participants will fix the rest of missing stuff =)
